cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/tms/TMSLogAppender.java

```diff
@@ -0,0 +1,267 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.logging.appender.tms;
+
+import ch.qos.logback.core.spi.ContextAware;
+import co.cask.cdap.api.messaging.MessagePublisher;
+import co.cask.cdap.api.messaging.MessagingContext;
+import co.cask.cdap.api.messaging.TopicNotFoundException;
+import co.cask.cdap.common.conf.CConfiguration;
+import co.cask.cdap.common.conf.Constants;
+import co.cask.cdap.common.service.RetryStrategies;
+import co.cask.cdap.common.service.RetryStrategy;
+import co.cask.cdap.logging.appender.LogAppender;
+import co.cask.cdap.logging.appender.LogMessage;
+import co.cask.cdap.logging.appender.kafka.LogPartitionType;
+import co.cask.cdap.logging.serialize.LoggingEventSerializer;
+import co.cask.cdap.messaging.MessagingService;
+import co.cask.cdap.messaging.context.MultiThreadMessagingContext;
+import co.cask.cdap.proto.id.NamespaceId;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Preconditions;
+import com.google.common.hash.Hashing;
+import com.google.common.util.concurrent.AbstractExecutionThreadService;
+import com.google.inject.Inject;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.Executor;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Log appender that publishes log messages to TMS.
+ */
+public final class TMSLogAppender extends LogAppender {
```

Comment:
This class has certain amount of duplicated code as the `KafkaLogAppender`. Please refactor it to have a common abstract class and abstract out the actual publish logic. Also, please use `AbstractRetryableScheduledService` instead of `AbstractExecutionThreadService` to simplify the logic.